server/lib/pbench/process_tb.py

```diff
@@ -76,61 +94,305 @@ def _results_push(controller: str, tb: Path, token: str):
+
+            # Combined operations will do
+            # 1. PUT API if have a token
+            # 2. p1
+            # 3. d1
+            # 4. d2
+            # 5. p3
+            # 6. p4
+            # 7. p5
+            # 8. p6
+            # 9. d3
+
+            # Extracts full path name of tar ball from the name of the .md5 file
+            # by trimming off the suffix.
             tb = Path(str(tbmd5)[0 : -len(".md5")])
-            tbdir = tb.parent
-            controller = tbdir.name
+            controller = tb.parent.name
 
-            try:
-                ProcessTb._results_push(controller, tb, self.token)
-            except Exception as e:
-                self.logger.error(
-                    "{}: Unexpected Error while running Agent's 'pbench-result-push' command: {}",
-                    self.config.TS,
-                    e,
-                )
-                nerr += 1
-                continue
+            if self.token:
+                # Step 1
+                # Only attempt to push a tar ball to a new Pbench Server via its
+                # PUT API when we have a token to use.  Any exceptions
+                # encountered are ignored s
+                try:
+                    ProcessTb._results_push(controller, tb, self.token)
+                except Exception as e:
+                    log_meth = (
+                        self.logger.warning
+                        if self.dispatch_states
+                        else self.logger.error
+                    )
+                    log_meth(
+                        "{}: Unexpected error while running Agent's 'pbench-result-push' command: {}",
+                        self.config.TS,
+                        e,
+                    )
+                    if not self.dispatch_states:
+                        # We have been asked to do one job, failures are errors
+                        # and we don't remove the failing tar ball so that the
+                        # PUT API can be retried.
+                        nerr += 1
+                        continue
 
-            try:
-                os.remove(tbmd5)
-            except Exception as exc:
-                self.logger.error(
-                    "{}: Cleanup of successful copy operation for MD5 file '{}' failed: '{}'",
-                    self.config.TS,
-                    tbmd5,
-                    exc,
-                )
-                nerr += 1
+            if not self.dispatch_states:
+                # We only have PUT API tasks cleanup the tar ball we just PUT.
+                nerr += self.remove(tbmd5, "cleanup of successful PUT API operation")
+                nerr += self.remove(tb, "cleanup of successful PUT API operation")
+            else:
+                assert False, f"unexpected {self.dispatch_states!r}"
+                # Step 2
+                # We have dispatch states configured, check to see if this tar
+                # ball is already in the archive tree.
```

Comment:
If we *also* had a `PUT` token, *and* we're sharing `ARCHIVE`, we expect `PUT` to have created the archive file, so we don't quarantine... we just go on to set the dispatch states.

Which actually makes me think we're trying to support four separate models, or at least don't have all the logic to completely define which *can* be supported:

1. Our planned beta deployment with shared `/srv/pbench` where
  a. put-shim does a PUT to the beta
  b. beta populates `ARCHIVE` and enables indexing (not backup or unpack)
  c. put-shim adds dispatch links
  d. put-shim backs up and unpacks for inventory reference
  e. beta indexes using temporary unpack
2. Two completely independent pipeline flows, where `PUT` starts a beta pipeline and `put-shim` goes on to unpack and index following the old 0.69 flow, with separate `/srv/pbench`. (That is, if we're not supporting this model, there's no point in retaining this quarantine logic.)
3. The earlier "put-shim" model where the put-shim server just does the `PUT` and then stops. (Which presumably would be marked by a `PUT` token and no dispatch states.)

The problem is that models 1 & 2 seem to both be represented here, but I don't think we have enough information to distinguish them. If we're not supporting 2, then this quarantine logic is obsolete and wrong. And, worse, I don't think we have enough information to distinguish between them. (That is, we need to know not only that we have a PUT token and dispatch states, but whether `ARCHIVE` is shared.)

I'd kinda thought the headset going into this, after yesterday, was that we were *only* supporting model 1 going forward. Adding additional compatibility will require extra knowledge.